src/Build/Graph/ProjectGraph.cs

```diff
@@ -219,47 +218,101 @@ public ProjectGraph(string entryProjectFile)
             return graphNode;
         }
 
+        private enum NodeState
+        {
+            // the project has been evaluated and it's project references are being processed
+            InProcess,
+            // all project references of this project have been processed
+            Processed
+        }
+
         /// <summary>
         /// Load a graph with root node at entryProjectFile
-        /// Maintain a queue of projects to be processed- each queue item is a key value pair of the project to be evaluated and its parent
-        /// Once the project has been evaluated, add a project reference to this evaluated target from the parent node
+        /// Maintain the state of each node (InProcess and Processed) to detect cycles
+        /// returns false if loading the graph is not successful
         /// </summary>
-        private void LoadGraph(string entryProjectFile, ProjectCollection projectCollection, PropertyDictionary<ProjectPropertyInstance> globalProperties)
+        private (bool success, List<string> projectsInCycle) LoadGraph(ConfigurationMetadata projectToEvaluate,
+            Dictionary<ConfigurationMetadata, NodeState> nodeState,
+            ProjectCollection projectCollection,
+            PropertyDictionary<ProjectPropertyInstance> globalProperties)
         {
-            var projectsToEvaluate = new Queue<KeyValuePair<ConfigurationMetadata, ProjectGraphNode>>();
-            // entry project node has no parent
-            projectsToEvaluate.Enqueue(new KeyValuePair<ConfigurationMetadata, ProjectGraphNode>(new ConfigurationMetadata(entryProjectFile, globalProperties), null));
-            while (projectsToEvaluate.Count != 0)
+            nodeState[projectToEvaluate] = NodeState.InProcess;
+            ProjectGraphNode parsedProject = CreateNewNode(projectToEvaluate, projectCollection);
+            IEnumerable<ProjectItemInstance> projectReferenceItems =
```

Comment:
What're your editor line wrap settings? It seems to be super aggressive with line wrapping.